### PR TITLE
Module#deprecate_constant with message

### PR DESCRIPTION
**Pull request by nobu**
Show the given message when a deprecated constant is referred.

### CODE REVIEW

File: test/ruby/test_module.rb [hunk 1]
```diff
@@ -1423,6 +1423,8 @@ def test_deprecate_constant
     assert_warn(/#{c}::FOO is deprecated/) {Class.new(c)::FOO}
     bug12382 = '[ruby-core:75505] [Bug #12382]'
     assert_warn(/deprecated/, bug12382) {c.class_eval "FOO"}
+    c.deprecate_constant(FOO: " by BAR")
```
Comment (znz) [inline test/ruby/test_module.rb]: I don't like separator space requires in string.

File: variable.c [hunk 2]
```diff
@@ -2228,12 +2228,25 @@ void
 rb_const_warn_if_deprecated(const rb_const_entry_t *ce, VALUE klass, ID id)
 {
     if (RB_CONST_DEPRECATED_P(ce)) {
-	if (klass == rb_cObject) {
-	    rb_warn("constant ::%"PRIsVALUE" is deprecated", QUOTE_ID(id));
+	if ((ce)->flag & CONST_DEPRECATED_MESG) {
+	    VALUE mesg = ((rb_deprecated_const_entry_t *)ce)->message;
+	    if (klass == rb_cObject) {
+		rb_warn("constant ::% "PRIsVALUE" is deprecated%"PRIsVALUE,
```
Comment (znz) [inline variable.c]: space after `%` intentionally?

File: variable.c [hunk 3]
```diff
@@ -2228,12 +2228,25 @@ void
...
+			rb_id2str(id), mesg);
```
Comment (znz) [inline variable.c]: Why do you change from `QUOTE_ID` to `rb_id2str`?

File: variable.c [hunk 4]
```diff
@@ -2228,12 +2228,25 @@ void
...
+	    }
+	    else {
+		rb_warn("constant %"PRIsVALUE"::% "PRIsVALUE" is deprecated%"PRIsVALUE,
+			rb_class_name(klass), rb_id2str(id), mesg);
```
Comment (znz) [inline variable.c]: ditto.

File: variable.c (hunk 2 above)
Comment (nobu) [inline variable.c]: Yes, space flag directs to quote unprintable characters.

File: variable.c (hunk 3 above)
Comment (nobu) [inline variable.c]: a space flag does.

**Pull request closed without merging**